### PR TITLE
Add composable filters to My Account order columns

### LINKED ISSUE #35635: Restructure hooks for My Orders table data
### Desired Solution
Remove the first action-based condition [here](https://github.com/woocommerce/woocommerce/blob/ba91c94ca9b1c4903964de70c8658cc7bff67d3f/plugins/woocommerce/templates/myaccount/orders.php#L44) and instead use **filters** to allow editing of each data element in the My Orders table. 

The existing action(s) could be maintained outside of this logic to ensure backwards compatibility. This solution would be consistent with WC & WP practices for customizing rendered data and would allow multiple modifications to be made without duplicating the default data.

### Describe alternatives you've considered

We are currently using snippets to override the behavior of individual plugins to address this, but it's clear that the underlying issue is with the template code itself.

### Current Behavior
The default My Orders table data is currently only rendered if no modifying actions are registered due to the first conditional check in the `if[...] elseif{` structure of [this section of the template](https://github.com/woocommerce/woocommerce/blob/ba91c94ca9b1c4903964de70c8658cc7bff67d3f/plugins/woocommerce/templates/myaccount/orders.php#L41-L77). 

### The Problem
This means that any dev/plugin wanting to add data here has to also render the default value themselves, causing duplications when multiple plugins are trying to modify the same data.

Here is an example with WooCommerce Gift Cards and Order Status Manager active, which both use `woocommerce_my_account_my_orders_column_order-status` to modify the Status value:
My account – svplugins 2022-11-17 at 8 09 49 PM

### PR DESCRIPTION
**Pull request by vladolaru**
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

The existing per-column actions are full-cell replacement hooks, which means extensions must reproduce the default markup and cannot safely compose changes from multiple callbacks. This adds a dynamic `woocommerce_account_orders_column_content_{$column_id}` filter that receives each column cell's full content — the escaped default HTML or, when callbacks are registered on the legacy `woocommerce_my_account_my_orders_column_{$column_id}` action, the output those callbacks produce — together with the current `WC_Order` and column ID, so extensions can amend the content instead of replacing it. This follows the same dual-hook pattern as `ReviewsListTable` (since 6.7.0) — an action for rendering custom columns plus a filter over the buffered cell output — with the `content` segment placed before the column ID rather than after it.

The legacy action keeps its replace-the-default semantics: when it has callbacks, the default markup is suppressed and the action output becomes the cell content. The filter then runs unconditionally over that content, so extensions can migrate from the action to the filter independently of each other — a filter callback keeps working on sites where another plugin still uses the action on the same column. The explicit `content_` segment avoids colliding with action-style `woocommerce_account_orders_column_{$column_id}` hooks already used in the ecosystem, because WordPress actions and filters share a callback registry.

Cell content is trimmed before filtering, so callbacks receive the bare markup — and an empty string, rather than template whitespace, for custom columns with no default renderer. String returns are used as-is; `int`, `float`, and `__toString`-capable object returns are coerced to strings; any other return (including `null`) triggers `wc_doing_it_wrong()` and falls back to the unfiltered content. Returning `''` renders an empty cell. Rows whose entry cannot be resolved to a `WC_Order` are skipped instead of fataling.

This is an additive public hook. No existing hook is renamed or removed, and the legacy action's arguments, timing, and default-suppressing behavior remain unchanged; the one new capability is that filter callbacks can post-process action-produced markup. The default output for valid orders is semantically unchanged aside from trimmed inter-tag whitespace inside table cells. Theme overrides of `myaccount/orders.php` need to reconcile the template version bump to 11.1.0 to receive the new filter. Until they do, filter callbacks never run on those sites, so an extension that moves from the legacy action to the filter silently stops contributing there. Extensions that need to cover outdated overrides should keep the legacy action *instead of*, not alongside, a filter callback emitting the same markup: when both are registered the action still suppresses the default cell content and the filter then runs over the action's output, rendering the addition twice. The change adds no site-scoped state and should behave identically on multisite, though multisite was not separately tested.

Default content remains escaped for its output context before filtering. As with other HTML filters, callbacks are responsible for returning safe HTML. The runtime cost is one in-memory output buffer and filter dispatch per order cell, with no new database, option, or HTTP work.

Closes #35635.

### How to test the changes in this Pull Request:

1. Create a customer account and at least one processing order for that customer.
2. Add these callbacks through a small test plugin or the active theme:

   ```php
   add_filter(
       'woocommerce_account_orders_column_content_order-status',
       function ( $content, $order, $column_id ) {
           return $content . '<span class="first-filter"> First filter</span>';
       },
       10,
       3
   );

   add_filter(
       'woocommerce_account_orders_column_content_order-status',
       function ( $content ) {
           return $content . '<span class="second-filter"> Second filter</span>';
       },
       20
   );
   ```

3. Sign in as the customer and open **My Account → Orders**.
4. Confirm the order status still renders once and is followed by both “First filter” and “Second filter” in priority order.
5. Add this legacy replacement callback:

   ```php
   add_action(
       'woocommerce_my_account_my_orders_column_order-status',
       function ( $order ) {
           echo '<span class="legacy-status">Legacy status</span>';
       }
   );
   ```

6. Reload **My Account → Orders** and confirm the status cell contains “Legacy status” followed by “First filter” and “Second filter”; the default status text should be absent.
7. Run the focused automated coverage:

   ```sh
   pnpm --filter=@woocommerce/plugin-woocommerce test:php:env -- --filter WC_My_Account_Orders_Template_Test
   ```

### Testing that has already taken place:

- Focused PHPUnit under PHP 8.1.34 (wp-env): 12 tests, 64 assertions passing — filter composition over default and legacy action output, custom columns, return-type coercion and rejection, whitespace trimming, and skipped unresolvable rows.
- Candidate-scoped PHPCS (`lint:php:changes`) and branch-scoped PHPCS (`lint:changes:branch`) against `origin/trunk`: passing.
- Empirical wp-env probes informed the design: buffered content for custom columns measured 18 bytes of raw template whitespace (now trimmed to `''`); `(string)` casts on non-stringable objects such as `WP_Error` fatal, ruling out blind coercion; an output-buffer-unbalancing callback corrupts rendering identically on trunk and on this branch, confirming the cell buffer adds no new exposure.
- Semgrep: 85 rules, zero findings.
- Changelog validation: passing.
- Independent review covered backward compatibility, extensibility, performance, output escaping, template overrides, and test cleanup.

### Milestone

- [x] Automatically assign milestone for the **[next WooCommerce version](../blob/trunk/plugins/woocommerce/woocommerce.php#L6)**

> **Note:** Check the box above to have the milestone automatically assigned when merged.
> Alternatively (e.g. for point releases), manually assign the appropriate milestone.

### Changelog entry

-   [ ] Automatically create a changelog entry from the details below.

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

-   [x] Patch
-   [ ] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [x] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message 

Add composable woocommerce_account_orders_column_content_{$column_id} filters for the My Account orders table columns and bump the myaccount/orders.php template version to 11.1.0.

</details>

### Release Communication

Select if this PR needs a generated summary for release notes:

- [ ] **Feature Highlight** - For user-facing features (what changed, user impact)
- [x] **Developer Advisory** - For developer-facing changes (what changed, how to detect, actions needed)

<details>
<summary>When to use each?</summary>

**Feature Highlight**: New features, UI changes, or improvements that merchants/store owners will notice.
- Example: "New bulk editing for products", "Improved checkout performance"

**Developer Advisory**: Breaking changes, deprecations, or changes that affect themes/plugins/extensions.
- Example: "Hook signature change", "Deprecated filter", "REST API field removed"

An AI will analyze your PR and post a draft comment for you to review and edit.
</details>

### Use of AI Tools

OpenAI Codex was used to independently review the implementation, run verification, and draft this Pull Request. Claude Code was used to evaluate review feedback, implement the current filter semantics, and verify them in the local environment. I reviewed the resulting code and description.

### CODE REVIEW

**Review by oaratovskyi — COMMENTED** [review]
Hey Vlad! Thanks for picking this one up. #35635 has been sitting open since 2022, and I like the solution. The one thing that stops me approving:

- **The new filter sits inside the `else` branch of the `has_action()` check**, so it never runs for a column any extension still hooks via the legacy action. The issue asks for the opposite in so many words: *"Remove the first action-based condition ... and instead use **filters** ... The existing action(s) could be maintained outside of this logic."* Gift Cards + Order Status Manager on `order-status` still duplicate exactly as before.
- **Core already ships the structure the issue asked for.** `ReviewsListTable::column_default()` (`src/Internal/Admin/ProductReviews/ReviewsListTable.php:1255-1292`, since 6.7.0) buffers its `do_action` and runs the result through a `..._content` filter *unconditionally*. Hoisting `ob_start()` above the `has_action()` check would match that precedent — and drop the ~36 lines of pure re-indentation this diff carries.
- **Mixed adoption fails silently.** Once one extension moves to the filter while another keeps the action, the migrated one's output disappears with no error. Nothing regresses against `trunk` today, but it's a trap built into a brand-new hook.

---

<details>
<summary><b>Detailed findings</b> — 1 important, 4 to address, 3 to consider</summary>

## Important (should address before merge)

### 1. The new filter never fires when the legacy action is registered — #35635's scenario is unchanged

`plugins/woocommerce/templates/myaccount/orders.php:54` (and `:114`)

`apply_filters( 'woocommerce_account_orders_column_content_' . $column_id, ... )` sits inside the `else` of `if ( has_action( 'woocommerce_my_account_my_orders_column_' . $column_id ) )` (`:54` → `:57` → `ob_start()` at `:58` → `ob_get_clean()` at `:99` → `apply_filters` at `:114` → `endif` at `:117`). When any callback is attached to the legacy action for a column, the entire buffered path — default markup *and* the new filter — is skipped for that column.

**The issue asks for exactly the change this declines to make.** #35635's body reads:

> Remove the first action-based condition [here] and instead use **filters** to allow editing of each data element in the My Orders table. The existing action(s) could be maintained outside of this logic to ensure backwards compatibility.

and names the collision: *"WooCommerce Gift Cards and Order Status Manager active, which both use `woocommerce_my_account_my_orders_column_order-status`."* Both callbacks still fire and both still render the default — the same visible duplication as before this PR. The composability win only arrives once *every* extension touching a column has migrated off the action, which is the one state that was never broken.

**Mixed adoption fails silently.** If Gift Cards moves to the new filter and Order Status Manager doesn't, Gift Cards' contribution vanishes — no error, no duplication, just missing content. Nothing regresses against `trunk` (the hook doesn't exist there yet), but whether a correctly-registered callback runs at all comes to depend on unrelated third-party plugin state, and nothing in the docblock (`:101-113`) warns about it.

Your test `test_legacy_order_status_column_action_still_replaces_default_output` (`class-wc-my-account-orders-template-test.php:93-117`) asserts this deliberately — `assertSame( 0, $filter_calls, 'Default-content filters should not run when a legacy action replaces the cell.' )` — so I'm reading it as an intentional BC decision rather than an oversight.

Two ways forward:

- **(a) Make the filter the single unconditional seam.** Hoist `ob_start()` above the `has_action()` check, keep the original `if`/`elseif` chain at its original indentation, and run whatever the branch produced through the filter. **Core already does this**: `ReviewsListTable::column_default()` (`src/Internal/Admin/ProductReviews/ReviewsListTable.php:1255-1292`) is `ob_start(); do_action( 'woocommerce_product_reviews_table_column_' . $column_name, $item ); echo $this->filter_column_output( ... );`, with `filter_column_output()` applying `..._content` unconditionally. So option (a) is the idiomatic shape here, not a novel one. It also preserves the action's replace-the-cell semantics exactly — only the *new* filter gains the ability to post-process — and it drops the ~36 lines of pure re-indentation this diff currently carries. The BC widening worth stating in the description is narrow: filter callbacks could then post-process action-produced markup.
- **(b) Keep the hard override, but be explicit.** Add one sentence to the docblock: *"This filter does not run when a callback is attached to `woocommerce_my_account_my_orders_column_{$column_id}`; it also lives only in this template, so it won't fire on sites with a stale `myaccount/orders.php` theme override."* Say it in the PR description too, and revisit `Closes #35635` — the issue's reported duplication is still there, so it isn't closed by this as written.

---

## Should address

### 2. Buffered content carries the template's own indentation

`orders.php:99`

`$column_content = (string) ob_get_clean();` captures everything between `ob_start()` (`:58`) and the `endif` (`:96`) — including the newlines and tab indentation that sit outside every conditional. This applies to **every** column, not just custom ones: each branch's output arrives wrapped in `"\n" + 8 tabs` on both sides. For a custom column registered via `wc_get_account_orders_columns()` that matches none of the five built-in branches, the buffer is 18 bytes of pure whitespace (`"\n\t\t\t\t\t\t\t\t\n\t\t\t\t\t\t\t\t"`) rather than `''`.

So any callback doing an exact match, `str_replace`, `strlen`, or `'' === $column_content` against the default gets a surprise, and the docblock ("Default column HTML") gives no hint. The existing tests use `assertStringContainsString`, which passes either way.

**Suggested:** `$column_content = trim( (string) ob_get_clean() );`. The only rendered difference is inter-tag whitespace inside table cells. Worth a test asserting the filter receives `''` for a registered custom column.

### 3. Changelog understates the change

`plugins/woocommerce/changelog/fix-35635-account-orders-column-filter`

Currently `Significance: patch` / `Type: tweak`. This PR adds a new public dynamic filter *and* bumps the template `@version` from `9.5.0` to `11.1.0`, which flags every theme override of `myaccount/orders.php` as outdated in WooCommerce → Status until the theme author reconciles it.

I couldn't find a written rule mandating the mapping, but the closest analogue in `changelog/` — `31365-add-backorder-notification-setting`, "a setting ... plus a `woocommerce_should_send_backorder_notification` filter" — is `minor`/`add`.

**Suggested:** regenerate as `Significance: minor` / `Type: add` (or `dev`), and mention the template version bump in the entry text so theme authors catch it in the release notes.

### 4. No test coverage for the new `continue` guard

`plugins/woocommerce/tests/php/includes/templates/class-wc-my-account-orders-template-test.php`

`orders.php:39-41` changes behavior from a fatal to a silent row skip. Every existing test builds `customer_orders->orders` from a valid `$order->get_id()` (`render_orders_template()`, test lines 259-273), so nothing exercises the guard. Since that helper already passes raw IDs, this is a small addition.

**Suggested:** add a case mixing a non-resolvable ID with a valid order ID, asserting the table renders with the valid row present and the invalid one skipped.

### 5. `is_string()` discards legitimate returns instead of coercing them

`orders.php:115`

Callbacks can reasonably return an `int`/`float` (natural for a numeric custom column), a `Stringable` object, or `null` to blank a cell. All three currently fall back to the stock markup with no notice. The `null` case is the sharpest: "return null to hide this cell" produces the opposite of what was asked. The tests cover `array` and `WP_Error` (test line 82) but none of these.

Worth noting the sibling precedent cuts the other way: `ReviewsListTable::filter_column_output()` returns `(string) apply_filters( ... )` — a coercion, not a gate.

**Suggested:** accept `string`/`int`/`float`/`Stringable` via `(string)`, treat `null` and `''` as an intentional empty cell, and reserve the fallback for genuinely non-stringable values — ideally paired with `wc_doing_it_wrong()` so the extension author learns their return was rejected.

---

## Consider

### 6. The filter is template-resident

`orders.php:114`

The `apply_filters()` call exists only in this template body, so a site running a theme-overridden `myaccount/orders.php` simply doesn't have the hook. The legacy action is less exposed — it's present in both `myaccount/orders.php` and the deprecated `myaccount/my-orders.php` (`:58-59`) and has been in overridden copies since 2.6. An extension author who migrates from action to filter regresses on exactly the sites where their action-based implementation worked, with no error. The `@version` bump is the right mitigation but is advisory only.

This is a WooCommerce-wide property, not something this PR introduces — the same is true of ~420 `do_action` and ~182 `apply_filters` calls across 228 template files — so I'm not asking for a structural fix. The cheap half is the docblock sentence already folded into finding #1's option (b).

If you *do* want the durable version some day: move the default per-column rendering into a function in `wc-account-functions.php` that applies the filter, so the template's job shrinks to calling it and the hook survives existing overrides.

### 7. The output buffer now wraps arbitrary third-party code

`orders.php:99`

`ob_start()` at `:58` wraps a region that runs third-party callbacks: `wc_get_account_orders_actions()` (`:80`) applies `woocommerce_my_account_my_orders_actions` (`includes/wc-account-functions.php:352`), and `get_formatted_order_total()` / `wc_format_datetime()` / `wc_get_order_status_name()` each run further filters. Two newly-possible failure modes:

- **A throw inside that region never reaches `ob_get_clean()`**, leaving the buffer open. Since this template renders inside `wc_get_template_html()`'s own buffer pair (`includes/wc-core-functions.php:369-373`), an unbalanced inner buffer corrupts the outer capture instead of failing cleanly.
- **A callback that unbalances buffering itself** (e.g. calls `ob_end_clean()`) destroys the inner buffer — and with `wc_get_template_html()`'s outer buffer open, line 99's `ob_get_clean()` then closes and returns the *outer* buffer's accumulated contents. Not `false`; `false` only happens on a front-end request with no ambient buffer. That means the whole surrounding page fragment can land inside one table cell, and a `false`-check wouldn't catch it.

I'd call this hardening rather than a demonstrated break — but I want to be careful not to overstate: first-party code *does* execute inside the buffered region (`includes/gateways/paypal/class-wc-gateway-paypal.php:215` hooks `woocommerce_my_account_my_orders_actions`), and I haven't traced every core callback for throw or buffer-unbalance behavior. There's also no in-repo precedent to lean on — `grouped.php:61` is the only other `ob_start()` in `templates/`, and it buffers `wc_get_template()`, not third-party hook callbacks.

**Suggested:** capture `ob_get_level()` before `ob_start()` and unwind to it in a `finally` — that's the remedy that catches both modes. Handling a `false` return is a secondary nicety for the no-ambient-buffer case.

### 8. Skipped rows have no empty state and no log entry

`orders.php:39`

The guard itself is right — keep it. But it removes the only signal (the fatal) that something went wrong, and opens two gaps:

- `$has_orders` and the table-vs-empty-state branch are decided before this loop, from `$customer_orders->total`. If every row fails the check, the page renders a full `<table>` with headers and an empty `<tbody>` instead of the "No order has been made yet." notice. A screen reader user lands on a table announced with columns and no rows, with no explanation.
- `total`/`max_num_pages` come from the paginated `wc_get_orders()` in `woocommerce_account_orders()` (`includes/wc-template-functions.php:3845-3855`), not from this loop — so a partial skip silently desyncs the rendered row count from what the pager advertises, with nothing in the log.

On reachability: with the default query only `WC_Order` comes back in practice (core has exactly two `WC_Abstract_Order` subclasses, and refunds aren't queried here), so the realistic triggers are a third-party `woocommerce_order_class` or `woocommerce_my_account_my_orders_query` filter, or a deleted/inaccessible order. Low severity, but cheap to make diagnosable.

**Suggested:** collect valid orders into an array before opening the `<table>` and gate the table on it being non-empty, falling through to the same `wc_print_notice()` empty state. Separately, log the skip via `wc_get_logger()->warning()` with the offending identifier before `continue`, so a merchant reporting missing rows produces evidence.

</details>

**Comment by vladolaru** [comment]
Hey @oaratovskyi — this is a genuinely strong review, and the central finding held up under every check I threw at it. I took each point through empirical validation in wp-env before deciding anything; here's where each one landed.

**1. Filter inside the `else` branch — accepted, implemented your option (a).** You're right that the mixed-adoption trap was real: the design punished exactly the extensions that migrate. `ob_start()` is now hoisted above the `has_action()` check, the original `if`/`elseif` chain is back at its trunk indentation, and the filter runs unconditionally over whatever the cell produced — I confirmed the `ReviewsListTable` shape you cited and mirrored it. One nuance worth noting: the issue's *literal* ask (defaults always render, actions maintained additively outside) would have made duplication worse — existing action callbacks render the default themselves, so core-default + each plugin's default-plus-extra would triple up. Keeping the action's replace-the-default semantics while making the filter the unconditional seam is the middle that resolves the issue without breaking anyone, and extensions can now migrate independently. The template diff vs trunk went from ±97 lines to +47/−2, as you predicted.

**2. Buffer whitespace — accepted.** I measured it: a custom column's buffer was exactly the 18 bytes of whitespace you predicted. Content is now `trim()`ed, and a test asserts custom-column filters receive `''`.

**3. Changelog — partially.** Reclassified as `Type: add` and the entry now names the filter and the template version bump. I kept `Significance: patch`, though: the `31365` analogue is no longer in the tree (cleaned on release), and the live precedent — `pr-64363`, which adds six new coupon validation filters — is `patch`/`add`, as are 9 of the 10 current `add` entries. Happy to bump it if you think the template-version impact tips it.

**4. Guard test — added.** A row with an unresolvable ID now renders alongside a valid one in the tests, asserting the skip.

**5. Return handling — partially.** `int`, `float`, and `__toString`-capable objects are now coerced, and any rejected return triggers `wc_doing_it_wrong()` so authors learn about it — both good calls. I kept `null` on the fallback path rather than treating it as an intentional blank: in WordPress filters, `null` overwhelmingly means a callback forgot to return, and silently blanking the cell would hide that bug; `''` is the explicit blank and works. On the precedent cutting the other way — it cuts back: I verified that `(string) new WP_Error()` throws a fatal `Error`, so `ReviewsListTable`'s blind coercion is the riskier half of that pattern. The doing-it-wrong message spells out the `''` contract.

**6. Template-resident filter — docblock updated.** It now states the filter only exists in the current template, so stale theme overrides won't have it. I'll leave the move into `wc-account-functions.php` for the "some day" you filed it under.

**7. Buffer hardening — I'm going to decline this one, with data.** I reproduced your scenario: a callback calling `ob_end_clean()` inside the region does corrupt the render (buffer level 1→0, outer buffer consumed). Then I ran the identical callback against *trunk's* template via the `wc_get_template` filter — same corruption, same level collapse, because `wc_get_template_html()`'s own buffer already wraps every third-party callback the template fires. The cell buffer adds a level but no new exposure. A `try`/`finally` here would guard one of the two buffer pairs while the outer one stays exposed, and there's no level-track-and-unwind precedent anywhere in `templates/` (the two `ob_get_level()` call sites in core serve pre-output buffer clearing). The throw case is moot in practice: nothing catches template render exceptions, so the request fatals with or without cleanup.

**8. Empty state and logging — declining, with reasoning.** Trunk's behavior for the same state is a fatal on `false->get_item_count()`, so the skip-and-render-remaining-rows behavior is strictly an improvement rather than a regression that needs compensating machinery. Reaching it requires a broken `woocommerce_order_class`/query filter or an order deleted mid-request, and restructuring the table flow in an overridable template makes every theme copy reconcile for that pathological state. On logging — no template in WooCommerce calls `wc_get_logger()` today, and I'd rather not have this template pioneer render-time logging. If the all-rows-skipped state ever shows up in the wild, the right home for both the detection and the empty-state decision is `woocommerce_account_orders()`, where the query runs.

The PR description is updated to state the final semantics, including the one genuinely new capability option (a) introduces: filter callbacks can post-process action-produced markup. Tests are at 12/64 assertions, all green.

**Comment by vladolaru** [comment]
@coderabbitai — on the outside-diff comment about `test_existing_account_order_column_actions_do_not_collide_with_content_filters`: that test isn't covering the legacy-hook contract, so the zero-call assertion is working as intended.

It guards the `content_` segment of the new filter name. WordPress keeps actions and filters in one `$wp_filter` registry, so if the filter were named `woocommerce_account_orders_column_{$column_id}` rather than `woocommerce_account_orders_column_content_{$column_id}`, `apply_filters()` would invoke third-party callbacks that plugins registered on that name with `add_action()` and take their return value — normally `null` — as the cell content. Registering an action on the bare name and asserting it never fires is exactly what breaks if that segment is ever dropped.

The legacy `woocommerce_my_account_my_orders_column_{$column_id}` contract is covered separately by `test_content_filters_compose_with_legacy_action_output()`, which registers the hook the template actually invokes and asserts both that it still suppresses the default markup and that content filters compose over its output.

The overlap you sensed was real, just elsewhere: that test and `test_custom_column_content_filter_receives_empty_string()` were building the same custom-column fixture and rendering the same template twice, so I merged them into one test that asserts all three properties from a single render. I also dropped the suite's hand-rolled hook registry and `tearDown()`, since `WP_UnitTestCase_Base` already restores the full `wp_filter` map after every test. The file is down from 355 to 273 lines with the same 64 assertions.

**Review by oaratovskyi — APPROVED** [review]
Hey Vlad 👋  

This one's in good shape, and thanks for taking the last round through `wp-env` before deciding anything. That reply was a pleasure to read. 

I did the same this time: ran both templates side by side on a real store (Storefront, HPOS, shortcode path), swapping only `orders.php`. Works well!

Pre-approving, nothing blocks merge. Everything below is a suggestion. I'd sort out at least finding 1 by polishing the doc block.

---

<details>
<summary><b>Details</b> — 3 suggestions</summary>

### 1. Call out the override trap in the release notes

`templates/myaccount/orders.php:120`

I dropped trunk's 9.5.0 template into Storefront and re-ran the two-extension case:

| | on a stale override |
|---|---|
| still on the legacy action | renders fine |
| migrated to your filter | **renders nothing** |

The docblock at `:110-112` already documents this, which is the right call. The bit that needs saying somewhere an extension author will actually read it (release notes, Developer Advisory) is: *additive, keep your action registration and add the filter alongside it.* That turns a trap into a migration path.

### 2. One line in the docblock about accessibility

`templates/myaccount/orders.php:100`

Three of the five default columns carry their only accessible affordances inside the HTML you hand to callbacks: the order-number link's `aria-label` (visible text is just `#123`), the action `aria-label`s, and `<time datetime>`. A replacing callback drops them silently. A sentence saying "preserve `aria-label` and `<time datetime>` if you replace rather than append" would cover it.

### 3. One small description fix

**The `ReviewsListTable` comparison isn't quite right.** That class uses `..._column_{$column}_content`; yours is `..._column_content_{$column}`. The naming is fine either way — `woocommerce_email_additional_content_{$id}` matches yours — it's only the "mirrors" claim that needs a word change.

---

**Closing out from last round.** Your decline on #8 stands, and my data backs it: no template in `templates/` calls `wc_get_logger()` or `error_log()` today, so you're right that this would be pioneering render-time logging — and since trunk fatals where your branch renders, there's no regression to compensate for. For the record the all-rows-skipped page does render a 5-column header with zero rows and a live "Previous" link; if that ever surfaces in the wild, `woocommerce_account_orders()` is the right home for it, exactly as you said. Also disregard my changelog note — I counted, `Type: add` runs 19 `minor` to 10 `patch` with public filters on both sides, so `patch` is fine as-is.

**Checked, no action needed:** the return-type ladder is correct all the way down — an object exposing `__toString` only via `__call` is properly *rejected*, since `method_exists()` doesn't see it. Your `null`-stays-on-the-fallback-path reasoning convinced me. And `trim()` does what it should: every column arrives clean, custom columns get exactly `''`.

</details>

**Comment by vladolaru** [comment]
Hey @oaratovskyi. All three are in. Two went in as you wrote them; on one I changed the guidance, and I want to flag why.

**2 — accessibility.** Added to the docblock. Your count checks out: `order-number` carries `aria-label="View order number %s"` with visible text of just `#123`, `order-date` carries `<time datetime>`, and `order-actions` carries a per-action `aria-label`. Three of five, and a replacing callback drops all of them silently.

**3 — the `ReviewsListTable` claim.** Correct, and reworded. `ReviewsListTable.php:1291` is `woocommerce_product_reviews_table_column_{$column_name}_content`; ours puts `content` before the ID. What genuinely does mirror is the pattern — the action at `:1268` for custom columns plus a filter over the buffered output — so the description now says that rather than claiming the shapes match. Since the hook is unreleased I could still rename it to the suffix form for an exact mirror; I'd rather not churn the template, eight tests and the changelog for it, but say the word if you'd prefer the consistency.

**1 — the override trap.** Took the gap, changed the wording. *Keep your action registration and add the filter alongside it* double-renders on up-to-date templates: `has_action()` is true, so the action's output **becomes** `$column_content`, and the filter then runs over it. `class-wc-my-account-orders-template-test.php:150-154` pins exactly that — the filter receives `legacy-status` and appends to it. An extension registering both hooks with the same markup emits it twice. Keeping the action also leaves the default cell suppressed on modern sites, which is both the thing this filter exists to fix and, per your finding 2, where the accessibility affordances live.

So the migration path reads as either/or rather than additive: filter-only going forward, or action-only where outdated overrides have to stay covered. That's now in the docblock beside the override note and in the PR description, so the Developer Advisory picks it up.

Docblock changes in cab5b4b91c and c6f7a3bd6c, description updated. If I've read the "additive" suggestion too literally and you had a both-registered pattern in mind that sidesteps the double render, I'm glad to swap it in.

File: plugins/woocommerce/tests/php/includes/templates/class-wc-my-account-orders-template-test.php [hunk 1]
```diff
@@ -0,0 +1,355 @@
...
+			'woocommerce_account_orders_column_content_order-type',
+			static function (): string {
+				return '<span class="filtered-order-type">Filtered order type</span>';
+			}
+		);
+
+		$html = $this->render_orders_template( $order );
+
+		$this->assertSame( 0, $legacy_action_calls, 'A content filter should not invoke callbacks registered on the existing action-style hook name.' );
+		$this->assertStringContainsString( 'Filtered order type', $html, 'A custom column should render content from its filter.' );
```
[the remark this answers is not in the document]
Comment (vladolaru) [reply plugins/woocommerce/tests/php/includes/templates/class-wc-my-account-orders-template-test.php:242]: Not taking this one — the test is doing what its name says, and the suggested change would remove the guard it provides.

Registering `woocommerce_account_orders_column_order-type` via `add_action()` works *because* the template never invokes it. WordPress keeps actions and filters in the same `$wp_filter` registry, so if this filter had been named `woocommerce_account_orders_column_{$column_id}` — the action-style name already used in the ecosystem — then `apply_filters()` would invoke that callback. The `assertSame( 0, ... )` is the regression guard on the `content` segment in the hook name, and it fails the moment someone renames the filter to the colliding form.

It isn't a vacuous assertion either: line 223 asserts the filter's own output rendered, so the test proves a conjunction — the content filter fired, and the same-prefix action did not.

Swapping in `woocommerce_my_account_my_orders_column_order-type` would instead assert that the legacy action runs once, which `test_content_filters_compose_with_legacy_action_output` already covers (it asserts the action output renders, the default is suppressed, and the filter receives that output).

One note on the committable suggestion regardless: it rewrites `++$legacy_action_calls;` as `+$legacy_action_calls;`, which is a no-op, so the proposed `assertSame( 1, ... )` would fail as written.

File: plugins/woocommerce/tests/php/includes/templates/class-wc-my-account-orders-template-test.php [hunk 2]
```diff
@@ -0,0 +1,355 @@
...
+			'woocommerce_account_orders_column_content_order-custom',
+			static function ( string $column_content ) use ( &$received_content ): string {
+				$received_content = $column_content;
+				return $column_content;
+			}
+		);
+
+		$this->render_orders_template( $order );
+
+		$this->assertSame( '', $received_content, 'A custom column with no default renderer should expose an empty string, not template whitespace.' );
```
[the remark this answers is not in the document]
Comment (vladolaru) [reply plugins/woocommerce/tests/php/includes/templates/class-wc-my-account-orders-template-test.php:269]: Added in `de03666e2d` as `test_empty_filtered_column_content_clears_default_output` — a filter on `order-status` returning `''`, asserting `Processing` is gone and the cell still renders.

The gap is real, though the stated reasoning isn't quite the mechanism. I mutated the template to `is_string( $filtered_column_content ) && '' !== $filtered_column_content` and ran the suite: that kills the new test, but it *also* fails the existing custom-column test, because falling through to the `else` branch emits a `wc_doing_it_wrong()` notice and `setExpectedIncorrectUsage()` isn't declared there. So a notice-emitting regression was already caught.

What wasn't covered is a **silent** fallback — for example a refactor to `$column_content = $filtered_column_content ?: $column_content;`. That passes the custom-column test (`'' ?: ''` is still `''`) and fails the new one. That's the boundary worth pinning, since `''` is valid content while `false` is not, and the two are one `empty()` apart.

Suite is green at 13 tests / 66 assertions against the current branch after the trunk merge.
